[Joel Tio] iP

joeltio opened the pull request:
# Ip Man

> I wish to fight ten men.
> — Ip Man

Ip Man offloads your mind of daily tasks so that you can be better at [Wing Chun](https://en.wikipedia.org/wiki/Wing_Chun). It's:

- 100% _Gluten-free_ and _organic_ 🌾
- Made only with ✨natural✨ ideas (no artificial LLM code)
- Good for your bones 🦴
- Very demure, very mindful 😇

To become a disciple, you just need to follow the soon-to-be-simplified process of:

1. Clone the repository from [here](https://github.com/joeltio/ip/tree/master)
2. Run the Gradle task `shawdowJar` with your preferred IDE
3. Run the produced Jar file (that can be found in `build/libs/ipMan.jar`) with `java -jar ipMan.jar`
4. Become Bruce Lee

Features:

- [x] Managing tasks (To-dos, Deadlines and Events)
- [x] Storing tasks even after program is closed
- [x] Find tasks by name
- [x] Mark tasks as done
- [ ] Many many more features to be revealed!

If you are a Java programmer and the white font in your dark-themed terminal hurts your eyes, consider overriding the `Ui` interface with your very own #3c4043 on #000000 themed UI!
```java
public interface Ui {
    // Inspired by https://nus-cs2103-ay2425s1.github.io/website/schedule/week3/project.html
    void showWelcome();
    Command readMessage();
    void showMessage(String message);
    void showMessageF(String message, Object... args);
    void showError(String message);
    void showGoodbye();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

joeltio committed:
Update Duke chatbot name to Ip Man

joeltio committed:
Update IpMan to echo user inputs and quit on "bye"

joeltio committed:
Update chatbot to record user inputs and list them when asked

joeltio committed:
Update chatbot to track inputs as tasks that can be marked done

joeltio committed:
Refactor command execution to separate classes

joeltio committed:
Refactor input parsing into command subclasses

joeltio committed:
Add specialized classes that extend Task

joeltio committed:
Add commands for creating tasks

joeltio committed:
Add tests for task types and marking done undone

joeltio committed:
Rename StringJoiner variables to Sj instead of Sb

joeltio committed:
Refactor extracting arguments from command to Parser

joeltio committed:
Refactor splitting input to command and args to Parser

joeltio committed:
Add exceptions for out of bounds index and missing args

joeltio committed:
Add delete command to remove items from list

joeltio committed:
refactor: move files into models and commands packages

joeltio committed:
refactor: rename Input record to Message

joeltio committed:
Add wrapper around NumberFormatException for better messages

joeltio committed:
Fix missing invalid index check for unmark command

joeltio committed:
Refactor and add tasks management and file io to TaskList

joeltio committed:
Refactor command classes to use TaskList

joeltio committed:
Replace TaskList with a file manager that saves after each command

joeltio committed:
Merge branch 'branch-Level-7'

joeltio committed:
Add support for LocalDate for Deadline

joeltio committed:
Add support for LocalDate for Event

joeltio committed:
Merge branch 'branch-Level-8'

joeltio committed:
Add find command

joeltio committed:
Merge branch 'branch-Level-9'

joeltio committed:
Merge branch 'add-gradle-support'

joeltio committed:
Rename duke in gradle to ipMan

joeltio committed:
Add checkstyle to gradle

joeltio committed:
Merge branch 'gradle-tasks'

joeltio committed:
Move parser package to root

joeltio committed:
Refactor throwing MissingArgumentException to only be done in Parser

joeltio committed:
Add ui abstraction layer and Cli implementation
Parser will be refactored later

joeltio committed:
Refactor Parser to parse raw messages strings and return commands

joeltio committed:
Refactor commands to use new Parser and add Context object

joeltio committed:
Refactor showing messages to Ui interface

joeltio committed:
Refactor list of tasks to TaskList

joeltio committed:
Merge branch 'A-MoreOOP'

joeltio committed:
Refactor all files to be under ipman package

joeltio committed:
Add tests for Parser

joeltio committed:
Add DummyFileManager and SpyUi for testing

joeltio committed:
Add tests delete command

joeltio committed:
Merge branch 'A-JUnit'

joeltio committed:
Add Javadoc and fix checkstyle issues for Parser

joeltio committed:
Add Javadoc and fix checkstyle issues for TaskList

joeltio committed:
Add Javadoc to TasksFileManager

joeltio committed:
Add Javadocs for Command and Context

joeltio committed:
Fix import order not conforming with coding standard

joeltio committed:
Merge branch 'A-CodingStandard'

cshao02 commented on src/main/java/ipman/commands/MarkCommand.java:
```diff
@@ -0,0 +1,25 @@
+import ipman.models.TaskList;
+
+public class MarkCommand implements Command {
+    private final int taskIndex;
```
Constants should be named in all caps?

joeltio replied:
Oh, in my opinion, constants which need to have all caps are constants that are inherently constant: things that will never change, things that are basically macros like `#DEFINE` in C. In other cases, I think it is alright to leave it as camel case.

Backing up my point, [the code convention](https://se-education.org/guides/conventions/java/intermediate.html) does not technically specify what constitutes a constant. But the [Google code style](https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names) which the code convention falls back on puts it more clearly:

> Constant names use `UPPER_SNAKE_CASE`: all uppercase letters, with each word separated from the next by a single underscore. But what is a constant, exactly?
>
> Constants are static final fields whose contents are deeply immutable and whose methods have no detectable side effects. Examples include primitives, strings, immutable value classes, and anything set to `null`. If any of the instance's observable state can change, it is not a constant. Merely intending to never mutate the object is not enough.

cshao02 commented on src/main/java/ipman/IpMan.java:
```diff
@@ -0,0 +1,38 @@
+
+
+public class IpMan {
+    private static final Ui ui = new CliUi();
+    private static final TaskList tasks = new TaskList();
+    private static final TasksFileManager fileManager = new TasksFileManager("./data/ipman.txt");
+    private static final Context context = new Context(tasks, ui, fileManager);
```
I noticed the same issue in several places too, but perhaps constants should be named in all caps?

joeltio replied:
I hold the same reasons for keeping these fields camel case as above too

cshao02 commented on src/main/java/ipman/ui/CliUi.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @Override
+    public void showMessageF(String message, Object... args) {
```
Perhaps a more descriptive name can be used here as it is not clear what the letter F means

joeltio replied:
I was a bit conflicted on how I should name this method. On one hand, the code convention tells me that I need to use camel case. On the other, nobody has ever named a formatting print function `printF`. This method is supposed to be the `printf` of displaying messages.

What do you think?

cshao02 replied:
Perhaps you could use showMessageFormatted or displayMessage, anything that can convey the idea of what the method is doing sufficiently

joeltio replied:
Resolved by https://github.com/nus-cs2103-AY2425S1/ip/pull/538/commits/338c4f7e995a96a70ec07341c163301fa9ab3e75

cshao02 commented on src/test/java/ipman/commands/DeleteCommandTest.java:
```diff
@@ -0,0 +1,38 @@
+import ipman.models.TaskList;
+import ipman.models.ToDo;
+import ipman.ui.SpyUi;
+import org.junit.jupiter.api.Test;
```
Perhaps similar import statements should be grouped together?

joeltio replied:
You're right, I realised my checkstyle did not run on test classes. I have corrected this.

cshao02 commented on src/test/java/ipman/ui/SpyUi.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.List;
+
+public class SpyUi implements Ui {
+    private final List<String> record = new ArrayList<>();
```
Perhaps plural form should be used for collection of items

joeltio replied:
Hmm, I think this field is a bit ambiguous. I was thinking of record as a collective noun i.e. I keep one record of all the messages shown. I have renamed this field to `recordedMessages` instead of `records`.

cshao02 reviewed (commented):
LGTM. Just a few nits to fix.

joeltio committed:
Add JavaDocs for DeleteCommand

joeltio committed:
Update InvalidIndexException to extend IndexOutOfBoundsException
Add JavaDocs for InvalidIndexException

joeltio committed:
Fix DeleteCommand to throw when index is invalid

joeltio committed:
Add JavaDocs to commands

joeltio committed:
Add JavaDocs to models

joeltio committed:
Add JavaDocs for parser and ui packages

joeltio committed:
Merge branch 'A-JavaDoc'

joeltio committed:
Fix checkstyle issues with tests
Add JavaDocs to test classes

joeltio committed:
Rename record to recordedMessages in SpyUi

joeltio commented:
Thanks for taking the time to review my PR @cshao02. I appreciate that you do so even though I have been committing a lot the past few days and doing so on top of your assigned PR.

joeltio committed:
Convert parseArgs to use varargs

marraknas commented on src/main/java/ipman/IpMan.java:
```diff
@@ -0,0 +1,41 @@
+package ipman;
+
+import ipman.commands.Command;
```
I think it's good that you  put the import statements in lexicographical order which is a common style error that people make!

marraknas commented on src/main/java/ipman/commands/CreateToDoCommand.java:
```diff
@@ -0,0 +1,38 @@
+
+    /**
+     * Creates command that creates a particular to-do
+     * @param name to-do name
```
Should there be a whitespace between the first line of javadocs and the next line? I mean this was repeated everywhere in the code so I was just wondering.

joeltio replied:
Hmm I think it's not a big deal, but I think it's a good idea to be consistent and have a newline before the [at]s

joeltio replied:
Resolved by https://github.com/nus-cs2103-AY2425S1/ip/pull/538/commits/38439567a10a1f9d6c06742e548b97b801ce2548

marraknas commented on src/main/java/ipman/commands/InvalidIndexException.java:
```diff
@@ -0,0 +1,24 @@
+     */
+    public InvalidIndexException(int capacity, int actual) {
+        super(String.format(
+            "Invalid index for list. Expected a number from 1 to %d, got %d",
```
A lot of people miss this and put the comma in the next line! Good that you did it right 😁

marraknas commented on src/main/java/ipman/models/Task.java:
```diff
@@ -0,0 +1,46 @@
+ * Represents a task that has a name and may or may not be completed.
+ */
+public abstract class Task {
+    protected boolean isDone = false;
```
just curious, how did you decide on these access modifiers?

joeltio replied:
In this case, I consider these fields to belong mainly to the classes that inherit this Task class (since Task can't be instantiated anyways). I think that it should be "private" to the subclass i.e. protected.

On the other hand, I could very well just expose `isDone` with a getter and disallow subclasses from modifying it. I'll update this to private and implement getters so that testing is easier. I will be able to check if a Task is marked by some command if I had a getter anyways.

joeltio replied:
Resolved by https://github.com/nus-cs2103-AY2425S1/ip/pull/538/commits/abf860ce968af86f71947ba360d1db9783c63039

marraknas commented on src/main/java/ipman/ui/CliUi.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @Override
+    public void showMessageF(String message, Object... args) {
```
Agreed that a better method name could be used in this case!

joeltio replied:
Resolved by https://github.com/nus-cs2103-AY2425S1/ip/pull/538/commits/338c4f7e995a96a70ec07341c163301fa9ab3e75

marraknas commented on src/test/java/ipman/parser/ParserTest.java:
```diff
@@ -0,0 +1,130 @@
+
+import org.junit.jupiter.api.Test;
+
+public class ParserTest {
```
A good amount of tests! Very comprehensive 🥳

marraknas reviewed (commented):
lgtm! Code is well written and tests are very comprehensive. Maybe a few whitespaces and style issues but otherwise really good!

joeltio committed:
Add JavaFx template to project

joeltio committed:
Update IpMan to be usable as a utility class
Remove readMessage from the Ui interface

joeltio committed:
Add MainWindow and view model to manage actions

joeltio committed:
Fix layout for MainWindow placement of nodes

joeltio committed:
Add MessageBox which displays user input and system reply

joeltio committed:
Update IpMan to only show message once per command

joeltio committed:
Update window to clear input after sending message

joeltio committed:
Remove showWelcome and showGoodbye from Ui
Update MainWindow to show welcome

joeltio committed:
Fix MessageBox responsiveness to resize and text alignment

joeltio committed:
Merge branch 'Level-10'

joeltio committed:
Add and enforce checkstyle rule for newline before Javadoc block tags

joeltio committed:
Add JavaDocs to Javafx code and fix other style issues

joeltio committed:
Rename showMessage to showMessageFormat

joeltio committed:
Update access modifiers for Task isDone and name

joeltio committed:
Add assertions to JavaFx components

JavaFx relies heavily on the existence of resource files or IDs of
particular nodes to exist in the FXML. When these are missing, this is
most certainly a programming error. Therefore, add assertions to ensure
they are properly loaded.

joeltio committed:
Refactor MessageBox to show system message by default

joeltio committed:
Refactor operations on observable messages in viewmodel

joeltio committed:
Add CI to check PRs and any pushes

Add a CI check to check whether the new branch that will be merged
conforms to checkstyle and passes the existing tests.

This is so that contributors can see almost immediately that their code
passes or fails the basic checks and can amend them promptly.

joeltio committed:
Fix missing runs-on for Github action

joeltio committed:
Merge pull request #1 from joeltio/branch-A-Assertions

Add assertions to JavaFx components

joeltio committed:
Merge branch 'master' into branch-A-CodeQuality

joeltio committed:
Merge pull request #2 from joeltio/branch-A-CodeQuality

Improve code quality

joeltio committed:
Merge pull request #3 from joeltio/branch-A-CI

Add CI for checking project

joeltio committed:
Remove unused import ListChangeListener

joeltio committed:
Add shorthand commands

joeltio committed:
Merge branch 'branch-BCD-Extension'

joeltio committed:
Update images to use creative commons licensed images

joeltio committed:
Update textfield to send message on enter

joeltio committed:
Add title, and width and height constraints to stage

joeltio committed:
Fix delete command calculates wrong index

joeltio committed:
Add message if there are no tasks to list

joeltio committed:
Shorten MainWindow comment to fit within 120 characters

joeltio committed:
Add user guide and Ui image

joeltio committed:
Add recovery mechanism if data file is not found
